### DISCUSSION TITLE
Level Generierung

### DISCUSSION THREAD

**Discussion by AMatutat** [Programmiermethoden/Dungeon#19] [Announcements]
Wir haben schon ein paar mal besprochen, dass es für unsere Rätsel gut wäre, wenn wir Level aus Graphen generieren könnten.

In meiner [Master-Arbeit](https://github.com/AMatutat/Prozedurale-graphenbasierte-2D-Level-Generierung) habe ich das ganze schonmal umgesetzt. 
Prinzipiell funktioniert das ganze auch, ist aber sehr sehr langsam. Liegt vor allem am Backtracking. 

Die Generierung der Level aus einem Baum hingegen sollte relativ flott gehen. Dafür muss der Algorithmus natürlich aufgeräumt werden.
Außerdem muss das Ganze an das neue Dungeon angepasst werden. 

Ich muss mir noch genau anschauen, WAS alles verändert/neu gemacht werden muss. 

Gibt es ggf. alternativen die schneller Einsatzbereit sind? 
Haben wir noch weitere Anforderungen an unseren Generator?

**Comment by malt-r** [comment]
Es ist in verschiedenen Diskussionen immer wieder die Idee von Teleporter-Türen aufgekommen (bspw. [hier](https://github.com/Programmiermethoden/Dungeon/discussions/6#discussioncomment-3813388)). Das ist auch relevant für die Levelgenerierung.

**Reply by malt-r** [reply]
## Räume geometrisch verbinden vs. Teleporter zwischen Räumen ##
  - geometrisch verbundene Räume sind immersiver, durch zusammenhängende Räume 
  wird Orientierungssinn besser gewahrt
  - isolierte Räume sind einfacher zu setzen, da nicht die geometrische Platzierung 
  beachtet werden muss
    - Generator muss dann nicht von jeder "falschen" Tür einen Weg zu einem Kampfraum 
      ziehen, oder x unterschiedliche Kampfräume platzieren (bezogen auf diese [Aufgabe](https://github.com/Programmiermethoden/Dungeon/discussions/6#discussioncomment-3697049))
    - setzt Definition von Übergängen voraus, wo soll das passieren?
      - Dot-Like Sprache in DSL zur direkten Definition einbetten?
      - wie kann das dann passend und ausreichend randomisiert werden?
- Mischweg:
  - Verbindung von Arealen/Sublevels durch Teleporter-Türen
  - ermöglicht, Räume mit Lehrinhalten und Erkundungsräume in einem Level zu verbinden
  - Erkundungsräume sollten generischer Content sein, der nicht explizit durch Lehrperson definiert 
    werden muss
  - Um Persistenz zu gewährleisten muss vermtl. viel/alles ahead of time generiert werden (zumindest)
    die Definition der Level
    - ist das direkt mit der Platzierung von Tiles & Entities verknüpft oder kann 
    das getrennt ablaufen (ressourcenschonender)?

**Reply by AMatutat** [reply]
Der Vorteile an diesen "Singe Room"-Lösungen (so nenn ich das jetzt einfach mal) ist, dass diese recht leicht in der Implementierung sind und wir zusätzlich keine  wirklichen Einschränkungen für unsere Graphen haben. 

Jeder Raum braucht so viele Türen, wie der Knoten Kanten im Graphen hat, eine Positionierung gibt es in dem Sinne nicht. 
Natürlich sollte eine gewisse Konsistenz erhalten bleiben, gehe ich in Raum A durch eine Tür oben in Raum B, sollte ich durch eine Tür unten in Raum B wieder nach Raum A gelangen.

**Reply by AMatutat** [reply]
Grundsätzlich braucht unser Level-System aber dann ein etwas größeres Rework. Aktuell ist das Framework nicht für diese Art von Level ausgelegt, sondern nur für "einfache" Levelstrukturen die als 2D-Array abgebildet werden können.

Ist aber nicht unmöglich das zu ändern.

**Reply by AlexJ2017** [reply]
## 3D Dungeon (Isometrisch vs. "Fake-Schrägperspektive")
Nach dem ersten Schritt der Generierung liegt ein 2D Array des Levels vor, in dem die begehbaren Flächen sowie leere Felder (nicht Teil des Dungeons) eingetragen sind.
Um dem Dungeon mehr Tiefe zu verleihen, stand eine Veränderung der Ansicht auf die Spielfläche im Raum.
### Isometrisch
Zum einen könnte das Spiel in einer **Isometrischen Perspektive** angezeigt werden, wie in diesem [Beispiel](http://1.bp.blogspot.com/-xq6sKUEfPcI/U0ebdoT8sLI/AAAAAAAAAIE/lrGWv_l4XIo/w1200-h630-p-k-no-nu/dark_pixels.gif).
Vorteile:
- relativ realistische 3D Ansicht der Level

Nachteile:
- Steuerung des Helden erfolgt entweder _gedreht_ und damit parallel zu dem Tile-Grid (Taste rechts -> rechts unten, hoch, Taste hoch -> rechts oben, Taste links-> links oben, hoch, Taste runter -> links unten)
oder Held bewegt sich _schräg_ durch die Level
- Texturen für die Helden, Gegner, Objekte ... müssten in isometrischer Form vorliegen, da sie sonst sehr flach wirken.
- Wände im Vordergrund können Teile des Spielfeldes verdecken und müssen evtl. transparent dargestellt werden

### Schrägperspektive
Alternativ kann das Level auch in einer Art Schrägperspektive angezeigt werden, in der nur die Wände des Dungeons perspektivisch dargestellt werden. Dadurch entsteht ein Gefühl von Tiefe, ohne großen Mehraufwand in der Darstellung der Level.
Es gibt verschiedene Möglichkeiten, diese Schrägperspektive umzusetzen. Eine Möglichkeit ist nur die hinteren Wände und die Türen des Levels darzustellen wie in diesem [Beispiel](https://upload.wikimedia.org/wikipedia/commons/9/95/Crawl_screenshot_6.png). Eine andere Möglichkeit wäre die Darstellung aller Wände in Schrägansicht ([Beispiel](https://www.keyforsteam.de/wp-content/uploads/the-binding-of-isaac-rebirth-800x500-1.jpg)).

![DungeonWalls](https://user-images.githubusercontent.com/29783124/195096645-8c551d28-b0eb-432b-a7cf-ef68f3faf5e5.png)

Die Wandtexturen müssten für die beiden unteren Möglichkeiten nur rotiert werden. Außerdem werden Texturen für die Ecken, Löcher und oder Säulen benötigt. Es sind außerdem schmalere Trennwände denkbar, welche statt der Löcher innerhalb der Räume eingesetzt werden könnten. Deren Darstellung könnte allerdings schwieriger sein, da viele Sonderfälle betrachtet werden müssen (mehr Texturen).

**Reply by AMatutat** [reply]
Die Isometrische Ansicht gefällt mir zwar optisch am besten, aber durch die aufgeführten Nachteile (und Sorge das sich dadurch noch mehr Probleme ergeben) bin ich für die  Schrägperspektive wie im zweiten Beispiel. 
Das scheint mir in der Umsetzung die wenigsten Fallstricke zu liefern, sollte  gut mit unserem aktuellen System kompatibel sein und zusätzlich sieht es auch noch recht modern aus.

**Reply by AlexJ2017** [reply]
![DungeonWallsTextured](https://user-images.githubusercontent.com/29783124/195323954-fd72212d-70dd-4d68-8b74-57c4d7e095eb.png)
Ich habe mal das obige Beispiel mit den Texturen vom alten Dungeon nachgebaut, um zu sehen wie es aussieht.

**Reply by AMatutat** [reply]
unten links find ich spitze

**Reply by AlexJ2017** [reply]
Nachträglich können dann ja einige Wände noch durch Türen ersetzt werden, je nach Bedarf.

**Comment by AMatutat** [comment]
# Single-Room Graphen Ansatz 

Um diese Diskussion mal in eine konkretes Gerüst zu Giesen.

## Konzept

Für viele unsere Ideen brauchen wir ein Graphenbasierte Levelstruktur (vgl. https://github.com/Programmiermethoden/Dungeon/discussions/6#discussioncomment-3697049 und https://github.com/Programmiermethoden/Dungeon/discussions/6#discussioncomment-3709068) 

Wie in/unter https://github.com/Programmiermethoden/Dungeon/discussions/19#discussioncomment-3813802 besprochen, ist eine *Single-Room Lösung (brauchen wir ein Glosar?) zwar nicht die immersivste Lösung aber die einfachste in der Implementierung und zusätzlich erlaubt uns das einiges an Freiheiten. 

Daher wäre die Idee einen Graphen zu generieren (entweder zufällig, oder per DSL Eingabe). Jeder Knoten ist ein Raum der generiert wird und jede Kante eine Tür zu einen nächsten Raum.

## Graphen Generierung

Die Level-Graphen würde ich mit einer DOT-Ähnlichen Sprache beschreiben. Per DSL kann dann der genaue Graph übergeben werden, alternativ (quasi wenn die Struktur des Levels keine Rolle spielt) generieren wir eine Beschreibung. 

Egal ob die Beschreibung per DSL oder Generator erzeugt wurde, aus dieser machen wir dann einen. Graphen.

**Anforderungen**: 
- Beschreibungssprache für die Graphen
- Beschreibungsgenerator
- Umwandeln der Beschreibung in einen Graphen

## Template/Raum Generierung

Für die Räume könnten wir Templates verwenden. Jedes Template gibt dabei die Größe und Form des Raums vor. Im Template selber könnte Platz für Mutation sein. In meiner Masterarbeit habe ich das so gemacht. Im Template gab es Wildcard-Flächen. Diese konnten dann durch Wildcard-Templates ersetzt werden. So hat man etwas mehr Abwechslung ohne sich die Mühe zu machen und viele Assets erstellt. 

**Anforderungen**
- Anzahl der Türen (und Position der Türen?) muss konfigurierbar sein. 
- Abwechslungsreich (evtl. ist das aber eher was für später und für den Anfang reicht uns eine Handvoll "Hardcoded" Räume) 

## Anbindung an das Dungeon

Hier gibt es einige Fallstricke denen ich mich langsam versuche zu stellen. 
Meine Grundidee ist es, jeden Raum als `TileLevel` Objekt zu betrachten. Also so wie wir jetzt ein normales Level betrachten. 
Ein `GraphLevel` ist dann eine Liste mit `TileLevel` und die Verbindungen dazwischen. 

Probleme vor denen ich noch stehe
- Wie bindet man den `GraphLevel` an das ÌLevel` bzw `LevelAPI` an. Wir benötigen spezielle Abfrag-Methoden, welche andere Level nicht benötigen. (Vielleicht Not.Supported Exception und die Graphen Methoden auch ins Interface?)
- Wie stellt man am besten Sicher, dass die Türen in sich logisch sind. Geh ich von Raum A nach unten in Raum B, muss ich wenn ich in Raum B nach oben gehe wieder in Raum A landen

### Pathfinding
Das libGDX Pathfinding, welches wir verwenden, geht immer von Tile zu Tile. Es sollte nichts dagegen sprechen, das Tür-Tile aus Raum A mit dem Tür-Tile aus Raum B zu verbinden. Das Pathfinding sollte weiter funktionieren. 


tbd: Skizze 

## Perspektive

https://github.com/Programmiermethoden/Dungeon/discussions/19#discussioncomment-3850881
Die Art und weise wie wir auf das Dungeon schauen, sollte für die Generierung egal sein, oder?
